libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/idp/IDPWebViewActivity.java

```diff
@@ -0,0 +1,171 @@
+            new RefreshAuthTokenTask(idpRequestHandler, webView).execute();
+        } catch (IDPRequestHandler.IDPRequestHandlerException e) {
+            SalesforceSDKLogger.e(TAG, "Building IDP request handler failed", e);
+            handleError("Incomplete SP config or user account data");
+        }
+    }
+
+    /**
+     * This class is used to monitor redirects within the WebView to determine
+     * when the login flow is complete, parses the response and passes it back.
+     */
+    protected class IDPWebViewClient extends WebViewClient {
+
+        @Override
+        public boolean shouldOverrideUrlLoading(WebView view, String url) {
+            boolean isDone = url.replace("///", "/").toLowerCase(Locale.US).startsWith(spConfig.getOauthCallbackUrl().
+                    replace("///", "/").toLowerCase(Locale.US));
+            if (isDone) {
+                final Uri callbackUri = Uri.parse(url);
+                final Map<String, String> params = UriFragmentParser.parse(callbackUri);
+
+                // Determines if the authentication flow succeeded or failed.
+                if (params != null) {
+                    final String code = params.get(CODE_KEY);
+                    if (TextUtils.isEmpty(code)) {
+                        handleError("Code not returned from server");
+                    } else {
+                        handleSuccess(code);
+                    }
+                } else {
+                    handleError("Code not returned from server");
+                }
+            }
+            return isDone;
+        }
+    }
+
+    private void handleError(String error) {
+        final Intent intent = new Intent();
+        intent.putExtra(ERROR_KEY, error);
+        setResult(RESULT_CANCELED, intent);
+        finish();
+    }
+
+    private void handleSuccess(String code) {
+        final Intent intent = new Intent();
+        intent.putExtra(CODE_KEY, code);
+        setResult(RESULT_OK, intent);
+        finish();
+    }
+
+    private class RefreshAuthTokenTask extends AsyncTask<Void, Void, String> {
+
+        private IDPRequestHandler idpRequestHandler;
+        private WebView webView;
+
+        public RefreshAuthTokenTask(IDPRequestHandler idpRequestHandler, WebView webView) {
+            this.idpRequestHandler = idpRequestHandler;
+            this.webView = webView;
+        }
+
+        @Override
+        protected String doInBackground(Void... nothings) {
+            String accessToken = null;
+            try {
+                accessToken = idpRequestHandler.getValidAccessToken();
+            } catch (Exception e) {
+                SalesforceSDKLogger.e(TAG, "Refreshing token failed", e);
+            }
+            return accessToken;
+        }
+
+        @Override
+        protected void onPostExecute(String accessToken) {
+            try {
+                if (accessToken != null) {
```

Comment:
Once the network request is done, we load frontdoor in the `WebView` -> this is done on the UI thread.